src/StoreApi/Utilities/OrderController.php

```diff
@@ -431,15 +431,54 @@ protected function validate_coupon_usage_limit( \WC_Coupon $coupon, \WC_Order $o
 		$coupon_usage_limit = $coupon->get_usage_limit_per_user();
 
 		if ( $coupon_usage_limit > 0 ) {
-			$data_store  = $coupon->get_data_store();
-			$usage_count = $order->get_customer_id() ? $data_store->get_usage_by_user_id( $coupon, $order->get_customer_id() ) : $data_store->get_usage_by_email( $coupon, $order->get_billing_email() );
+			$data_store = $coupon->get_data_store();
+			// First, we check a logged in customer usage count, which happens against their user id, billing email, and account email.
+			if ( get_current_user_id() ) {
+				// We get usage per user id and associated emails.
+				$usage_count = $this->get_usage_per_aliases( $coupon, array( $order->get_billing_email(), get_current_user_id(), wp_get_current_user()->user_email ) );
+			} else {
+				// Otherwise we check if the email doesn't belong to an existing user.
+				$customer_data_store = \WC_Data_Store::load( 'customer' );
+				// This will get us any user ids for this billing email.
+				$user_ids       = $customer_data_store->get_user_ids_for_billing_email( array( $order->get_billing_email() ) );
+				$emails_for_ids = array_map(
+					function( $user_id ) {
+						return get_userdata( $user_id )->user_email;
+					},
+					$user_ids
+				);
+				$usage_count    = $this->get_usage_per_aliases( $coupon, array_merge( $emails_for_ids, $user_ids, array( $order->get_billing_email() ) ) );
+			}
 
 			if ( $usage_count >= $coupon_usage_limit ) {
 				throw new Exception( $coupon->get_coupon_error( \WC_Coupon::E_WC_COUPON_USAGE_LIMIT_REACHED ) );
 			}
 		}
 	}
 
+	/**
+	 * Get the usage count for a coupon based on a list of aliases (ids, emails).
+	 *
+	 * @param \WC_Coupon $coupon Coupon object applied to the cart.
+	 * @param array      $aliases List of aliases to check.
+	 *
+	 * @return integer
+	 */
+	public function get_usage_per_aliases( &$coupon, $aliases ) {
+		global $wpdb;
+		$aliases        = array_unique( array_filter( $aliases ) );
+		$aliases_string = "('" . join( "','", array_map( 'esc_sql', $aliases ) ) . "')";
```

Comment:
```suggestion
		$aliases_string = "('" . implode( "','", array_map( 'esc_sql', $aliases ) ) . "')";
```

`join` is an alias. Implode is preferred.